### PR TITLE
[REVIEW] Using the new UCX-Py

### PR DESCRIPTION
**Pull request by madsbk**
This PR makes cudf use the new [UCX Python bindings](https://github.com/rapidsai/ucx-py/pull/180). 
It depend on this DASK PR: [#3059](https://github.com/dask/distributed/pull/3059), which has now been merged.

### CODE REVIEW

File: python/dask_cudf/record.txt [hunk 1]
```diff
@@ -1,65 +1,65 @@
-/conda/envs/gdf/lib/python3.7/site-packages/dask_cudf/core.py
```
Comment (quasiben) [inline python/dask_cudf/record.txt]: Not sure why `record.txt` is in the repo -- @kkraus14 should this file be added to .gitignore ?

Comment (kkraus14) [reply python/dask_cudf/record.txt]: Yes it should be.

Comment (madsbk) [reply python/dask_cudf/record.txt]: Ok, I will delete and ignore it

File: python/cudf/cudf/core/buffer.py [hunk 2]
```diff
@@ -110,7 +96,20 @@ def deserialize(cls, header, frames):
         obj : Buffer
             Returns an instance of Buffer.
         """
-        return Buffer(frames[0], header=header)
+        assert len(frames) == 1, "Use Buffer.serialize() for serialization"
+        iface = header["cuda_array_interface"]
+        if iface["shape"][0] > 0:
+            arr = numba.cuda.devicearray.DeviceNDArray(
```
Comment (shwina) [inline python/cudf/cudf/core/buffer.py]: Should this be an `rmm` backed device array instead?

Comment (quasiben) [reply python/cudf/cudf/core/buffer.py]: @shwina probably.  Is that `rmm.to_device` ?

Comment (madsbk) [reply python/cudf/cudf/core/buffer.py]: Good point, fixed

**Comment by madsbk** [comment]
@kkraus14, @shwina. If the CI passes, I think this PR is ready to be merged.

**Comment by quasiben** [comment]
With https://github.com/dask/distributed/pull/3059 in , I believe this PR is the last piece for UCX/cuDF/Dask and then we can start in earnest for testing/benchmarking

File: python/cudf/cudf/core/buffer.py (hunk 2 above)
Comment (shwina) [reply python/cudf/cudf/core/buffer.py]: Looks like it's still a numba devicearray?

File: python/cudf/cudf/core/buffer.py [hunk 3]
```diff
@@ -110,7 +96,21 @@ def deserialize(cls, header, frames):
...
+                iface["shape"],
+                iface["strides"],
+                np.dtype(iface["typestr"]),
+                gpu_data=numba.cuda.as_cuda_array(frames[0]).gpu_data,
+            )
+            arr, _ = rmm.auto_device(arr)
+        else:
+            arr = numba.cuda.device_array(
+                (0,), dtype=np.dtype(iface["typestr"])
+            )
```
Comment (kkraus14) [inline python/cudf/cudf/core/buffer.py]: Here too, should use rmm

**Comment by madsbk** [comment]
@kkraus14, @shwina, removed the use of `numba.cuda`.

File: python/cudf/cudf/core/buffer.py [hunk 4]
```diff
@@ -98,19 +90,28 @@ def deserialize(cls, header, frames):
...
-            arr, _ = rmm.auto_device(arr)
         else:
-            arr = numba.cuda.device_array(
-                (0,), dtype=np.dtype(iface["typestr"])
-            )
-        return Buffer(arr, header=header)
+            # Updating the data pointer to the frame data.
+            iface["data"] = frames[0].__cuda_array_interface__["data"]
+
+            class _dummy_iface_obj:
```
Comment (shwina) [inline python/cudf/cudf/core/buffer.py]: Could use a `SimpleNamespace` here instead:

https://docs.python.org/3/library/types.html#types.SimpleNamespace

**Comment by kkraus14** [comment]
rerun tests

**Comment by quasiben** [comment]
@shwina @kkraus14 i think @madsbk has addressed all the rmm and style issues, correct ?

**Pull request merged**